addons/udes_stock_refactoring/models/stock_move.py

```diff
@@ -0,0 +1,437 @@
+                    action = picking_type.u_post_validate_action
+                else:
+                    continue  # Don't refactor cancel or draft moves.
+
+                if action:
+                    _logger.info(
+                        "Refactoring %s at %s using %s: %s",
+                        picking_type.name,
+                        stage,
+                        action,
+                        st_moves.ids,
+                    )
+                    func = getattr(st_moves, "refactor_action_" + action)
+                    new_moves = func()
+                    if new_moves is not None:
+                        moves -= st_moves
+                        moves |= new_moves
+
+        return moves
+
+    def _action_confirm(self, *args, **kwargs):
+        """Extend _action_confirm to trigger refactor action.
+
+        Odoos move._action_confirm returns all the moves passed in to it, after
+        merging any it can. In places the return value is used to immediately
+        assign, so any created moves should be returned.
+
+        However, in stock.move._split() in Odoo core, _action_confirm is
+        called on a single move and it expects a single move to be returned
+        (will error out if not). Therefore the refactor at confirm should avoid
+        splitting/creating moves. Luckily this is unlikely to be an issue as
+        it makes little sense to split moves at this stage, before they have
+        stock reserved against them.
+        """
+        res = super(StockMove, self)._action_confirm(*args, **kwargs)
+        post_refactor_moves = res._action_refactor(stage="confirm")
+
+        if post_refactor_moves != res:
+            raise UserError(
+                _(
+                    "Post confirm refactor has created or destroyed "
+                    "moves, which could break things if you have the "
+                    "MRP module installed"
+                )
+            )
+        return res
+
+    def _action_assign(self):
+        """Extend _action_assign to trigger refactor action.
+        n.b. _action_assign does not return anything in core Odoo, so we
+        don't return any extra moves that may have been created
+        by refactoring.
+        """
+        res = super(StockMove, self)._action_assign()
+
+        self.exists()._action_refactor(stage="assign")
+        return res
+
+    def _action_done(self, cancel_backorder=False):
+        """Extend _action_done to trigger refactor action, and push from drop
+
+        Odoo returns completed moves.
+        Therefore we will keep track of moves created by the refactor and
+        return them as part of the set of completed moves.
+        """
+        done_moves = super(StockMove, self)._action_done()
+
+        post_refactor_done_moves = done_moves._action_refactor(stage="validate")
+
+        post_refactor_done_moves.push_from_drop()
+        return post_refactor_done_moves
+
+    def push_from_drop(self):
```

Comment:
This should go into another module.

Reply:
Into ```udes_stock``` or another stock addon module?

Reply:
A new module. But not needed for the refactoring code.

Reply:
This has now been removed.